reactor-netty-core/src/main/java/reactor/netty/transport/TransportConnector.java

```diff
@@ -252,56 +261,84 @@ static void doConnect(
 
 					int next = index + 1;
 					if (next < addresses.size()) {
-						connectPromise.setFailure(new RetryConnectException(addresses));
+						connectPromise.asPromise().setFailure(new RetryConnectException(addresses));
 					}
 					else {
-						connectPromise.setFailure(cause);
+						connectPromise.asPromise().setFailure(cause);
 					}
 				}
 			});
 		});
 	}
 
-	@SuppressWarnings("FutureReturnValueIgnored")
 	static Mono<Channel> doInitAndRegister(
 			TransportConfig config,
 			ChannelInitializer<Channel> channelInitializer,
 			boolean isDomainSocket,
 			EventLoop eventLoop,
 			ContextContainer container) {
-		ChannelFactory<? extends Channel> channelFactory = config.connectionFactory(config.eventLoopGroup(), isDomainSocket);
-
-		Channel channel = null;
+		boolean onServer = channelInitializer instanceof ServerTransport.AcceptorInitializer;
+		Channel channel;
 		try {
-			channel = channelFactory.newChannel();
-			container.save(channel);
-			if (channelInitializer instanceof ServerTransport.AcceptorInitializer) {
+			if (onServer) {
+				EventLoopGroup childEventLoopGroup = ((ServerTransportConfig<?>) config).childEventLoopGroup();
+				ServerChannelFactory<? extends Channel> channelFactory = config.serverConnectionFactory(isDomainSocket);
+				channel = channelFactory.newChannel(eventLoop, childEventLoopGroup);
 				((ServerTransport.AcceptorInitializer) channelInitializer).acceptor.enableAutoReadTask(channel);
 			}
-			channel.pipeline().addLast(channelInitializer);
-			setChannelOptions(channel, config.options, isDomainSocket);
-			setAttributes(channel, config.attrs);
+			else {
+				ChannelFactory<? extends Channel> channelFactory = config.connectionFactory(isDomainSocket);
+				channel = channelFactory.newChannel(eventLoop);
+			}
 		}
 		catch (Throwable t) {
-			if (channel != null) {
-				channel.unsafe().closeForcibly();
-			}
 			return Mono.error(t);
 		}
```

Comment:
why the channel is not closed anymore  in case of an exception ?

Reply:
I something fails it will be channel creation? Do we have a channel if there is an exception?

Reply:
oops, yeah, sorry.